src/osd/scrubber/osd_scrub_sched.cc

```diff
@@ -0,0 +1,551 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "osd_scrub_sched.h"
+
+#include "include/utime.h"
+#include "osd/OSD.h"
+
+#include "pg_scrubber.h"
+
+
+#define dout_context (cct)
+#define dout_subsys ceph_subsys_osd
+#undef dout_prefix
+#define dout_prefix *_dout << "osd." << whoami << "  "
+
+// ////////////////////////////////////////////////////////////////////////////////// //
+// ScrubJob
+
+ScrubQueue::ScrubJob::ScrubJob(CephContext* cct, const spg_t& pg, int node_id)
+    : RefCountedObject{cct}
+    , sched_time{}
+    , deadline{0, 0}
+    , pgid{pg}
+    , whoami{node_id}
+    , cct{cct}
+{}
+
+
+// ////////////////////////////////////////////////////////////////////////////////// //
+// ScrubQueue
+
+#undef dout_context
+#define dout_context (m_cct)
+#undef dout_prefix
+#define dout_prefix \
+  *_dout << "osd." << m_osds.whoami << " scrub-queue::" << __func__ << " "
+
+
+ScrubQueue::ScrubQueue(CephContext* cct, OSDService& osds) : m_cct{cct}, m_osds{osds}
+{
+  // initialize the daily loadavg with current 15min loadavg
+  if (double loadavgs[3]; getloadavg(loadavgs, 3) == 3) {
+    daily_loadavg = loadavgs[2];
+  } else {
+    derr << "OSD::init() : couldn't read loadavgs\n" << dendl;
+    daily_loadavg = 1.0;
+  }
+}
+
+std::optional<double> ScrubQueue::update_load_average()
+{
+  int hb_interval = m_cct->_conf->osd_heartbeat_interval;
+  int n_samples = 86400;
+  if (hb_interval > 1) {
+    n_samples /= hb_interval;
+    if (n_samples < 1)
+      n_samples = 1;
+  }
+
+  // get CPU load avg
+  double loadavgs[1];
+  if (getloadavg(loadavgs, 1) == 1) {
+    daily_loadavg = (daily_loadavg * (n_samples - 1) + loadavgs[0]) / n_samples;
+    dout(17) << "heartbeat: daily_loadavg " << daily_loadavg << dendl;
+    return 100 * loadavgs[0];
+  }
+
+  return std::nullopt;
+}
+
+void ScrubQueue::remove_from_osd_queue(ceph::ref_t<ScrubJob> scrub_job)
+{
+  if (scrub_job->m_in_queue) {
+    dout(10) << " removing pg(" << scrub_job->pgid << ") from OSD scrub queue" << dendl;
+
+    scrub_job->sched_time = {};	 // appearing in logs/asok
+
+    const auto target_pg = scrub_job->pgid;
+    auto same_pg = [&target_pg](const auto& jobref) { return jobref->pgid == target_pg; };
+
+    // search for this entry in both queues
+    std::unique_lock l(jobs_lock, 200ms);
```

Comment:
Originally - yes. But now I am not sure it's a bad idea. A failure to release this lock
can only happen due to a bug in my code.
Two issues with that code remaining:
1 - the timed-lock used instead of a simple mutex is somewhat more expensive (not sure it's an issue here), and
2 - the timeout is just a guess. But a very long timeout (a few seconds, possibly) will serve here as well,
   to crush the OSD in the relevant spot.